components/calendar/src/chinese.rs

```diff
@@ -0,0 +1,1213 @@
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5318-L5327
+    fn chinese_new_moon_before(moment: Moment) -> RataDie {
+        let new_moon_moment = Astronomical::new_moon_before(Self::midnight_in_china(moment));
+        let chinese_offset = Self::chinese_offset(new_moon_moment.as_rata_die());
+        Location::standard_from_universal(new_moon_moment, chinese_offset).as_rata_die()
+    }
+
+    /// Universal time of midnight at start of a Moment in China
+    ///
+    /// Based on functions from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5353-L5357
+    fn midnight_in_china(date: Moment) -> Moment {
+        Location::universal_from_standard(date, Self::chinese_offset(date.as_rata_die()))
+    }
+
+    /// Determines the fixed date of the Chinese new year in the sui4 (solar year based on the winter solstice)
+    /// which contains the fixed date passed as an argument.
+    ///
+    /// Based on functions from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5370-L5394
+    fn chinese_new_year_in_sui(date: RataDie) -> RataDie {
+        let prior_solstice = Self::chinese_winter_solstice_on_or_before(date); // s1
+        let following_solstice = Self::chinese_winter_solstice_on_or_before(prior_solstice + 370); // s2
+        let month_after_eleventh =
+            Self::chinese_new_moon_on_or_after((prior_solstice + 1).as_moment()); // m12
+        let month_after_twelfth =
+            Self::chinese_new_moon_on_or_after((month_after_eleventh + 1).as_moment()); // m13
+        let next_eleventh_month =
+            Self::chinese_new_moon_before((following_solstice + 1).as_moment()); // next-m11
+        let m12_float = month_after_eleventh.as_moment().inner();
+        let next_m11_float = next_eleventh_month.as_moment().inner();
+        let lhs_argument =
+            libm::round((next_m11_float - m12_float) / astronomy::MEAN_SYNODIC_MONTH) as i64;
+        if lhs_argument == 12
+            && (Self::chinese_no_major_solar_term(month_after_eleventh)
+                || Self::chinese_no_major_solar_term(month_after_twelfth))
+        {
+            Self::chinese_new_moon_on_or_after((month_after_twelfth + 1).as_moment())
+        } else {
+            month_after_twelfth
+        }
+    }
+
+    /// Get the moment of the nearest winter solstice on or before a given fixed date
+    ///
+    /// Based on functions from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5359-L5368
+    fn chinese_winter_solstice_on_or_before(date: RataDie) -> RataDie {
+        let approx = Astronomical::estimate_prior_solar_longitude(
+            270.0,
```

Comment:
suggestion: extract magic number as constant `WINTER`. perhaps put in astronomy module.